### PR TITLE
Add deadline banners & remove the apply button in preview mode

### PR DESCRIPTION
**Pull request by davidgisbey**
### Context

This PR does 2 things

1. Uses the CycleTimetable to automate Find moving between the 3 discreet phases 
2. Adds 3 different banners which should show based on which deadlines are upcoming and have passed. 

### Changes proposed in this pull request

- 1st banner shows from Monday 12/6 until the apply1 deadline 

image

- 2nd banner shows from the apply1 deadline until the apply 2 deadline 

image

- 3rd banner shows after apply2 has closed and before find closes 

image

### Guidance to review

Just adding in some feature tests to ensure it doesn't show on the course page mid cycle and that it doesn't show on the subjects filter pages period.

### Trello card

https://trello.com/c/3MhlVfCk/3591-%F0%9F%94%81-find-eoc-dev-phase-1-banners-on-find-to-notify-candidates-the-cycle-will-end

https://trello.com/c/gyYZwYjd/3631-%F0%9F%94%81-find-eoc-dev-phase-2-banners-on-find-notify-candidates-that-apply-1-cycle-has-ended-but-apply-2-is-still-open

https://trello.com/c/rFbcBD4o/3632-%F0%9F%94%81-find-eoc-dev-phase-3-banners-on-find-notify-candidates-that-both-apply-1-and-2-applications-are-closed-and-stop-showing-apply
### Checklist

- [x] Rebased master
- [x] Cleaned commit history
- [x] Tested by running locally

### CODE REVIEW

File: app/components/deadline_banner_component.html.erb [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+    <p class="govuk-body">Courses can fill up at any time, so you should apply as soon as you can.</p>
+    <p class="govuk-body">If you’re applying for the first time since applications opened in <%= CycleTimetable.find_opens.strftime('%B %Y') %>, apply no later than 6pm on <%= apply_1_deadline %>.</p>
+    <p class="govuk-body">If your application did not lead to a place and you’re applying again, apply no later than 6pm on <%= apply_2_deadline %>.</p>
+  <% end %>
+<% elsif CycleTimetable.show_apply_2_deadline_banner? %>
+  <%= govuk_notification_banner(title: "Important") do |notification_banner| %>
+    <%= notification_banner.slot(:heading, text: "If you’re applying for the first time since applications opened in #{CycleTimetable.find_opens.strftime('%B %Y')}") %>
+    <p class="govuk-body">It’s no longer possible to apply for teacher training starting in the <%= "#{CycleTimetable.current_year} to #{CycleTimetable.next_year}" %> academic year.</p>
+    <p class="govuk-body">Come back from 9am on <%= find_reopens %> to find courses starting in the <%= "#{CycleTimetable.next_year} to #{CycleTimetable.next_year + 1}" %> academic year.</p>
```
Comment (duncanjbrown) [inline app/components/deadline_banner_component.html.erb:12]: Shall we put these into the calendar as times?

File: app/views/layouts/application.html.erb [hunk 2]
```diff
@@ -120,6 +120,8 @@
             <%= notification_banner.slot(:heading, text: flash[:success]) %>
           <% end %>
         <% end %>
+        
+        <%= render DeadlineBannerComponent.new(flash_empty: flash.empty?) unless request.url.include?('/results/filter/subject')%>
```
Comment (duncanjbrown) [inline app/views/layouts/application.html.erb:124]: Maybe put this condition in a named method for explicitness?

File: config/routes.rb [hunk 3]
```diff
@@ -8,7 +8,7 @@
   get :healthcheck, controller: :heartbeat
   get :sha, controller: :heartbeat
 
-  if FeatureFlag.active?(:cycle_has_ended)
+  if CycleTimetable.find_down?
```
Comment (duncanjbrown) [inline config/routes.rb:11]: sorry, not in this PR, but `down` to me implies "broken", not "closed"?

File: app/views/courses/_apply.html.erb [hunk 4]
```diff
@@ -8,7 +8,7 @@
   <% if @course.has_vacancies? %>
     <%= render 'apply_button' %>
 
-    <% if FeatureFlag.active?(:display_apply_button) %>
+    <% if CycleTimetable.mid_cycle? %>
```
Comment (duncanjbrown) [inline app/views/courses/_apply.html.erb:11]: Vague desire to rename this method to `CycleTimetable.business_as_usual?`!?

File: app/components/deadline_banner_component.rb [hunk 5]
```diff
@@ -0,0 +1,29 @@
+class DeadlineBannerComponent < ViewComponent::Base
+  attr_accessor :flash_empty
+
+  def initialize(flash_empty:)
+    @flash_empty = flash_empty
+  end
+
+  def render?
+    flash_empty
```
Comment (duncanjbrown) [inline app/components/deadline_banner_component.rb:9]: It's a bit mixy to have this return true and then the component to render nothing?

Suggested refactor:

```

def render?
  flash_empty? && banner_to_show
end
  
private

def banner_to_show
  return :apply_1_deadline if CycleTimetable.show_apply_1_deadline_banner?
  return :apply_2_deadline if CycleTimetable.show_apply_2_deadline_banner?
  return :cycle_closed if CycleTimetable.show_cycle_closed_banner?
end
```

and then switch on `banner_to_show` in the template?

File: app/views/layouts/application.html.erb (hunk 2 above)
Comment (davidgisbey) [inline app/views/layouts/application.html.erb:124]: the cards say it shouldn't render on the subjects filter page

File: spec/features/courses/show_spec.rb [hunk 6]
```diff
@@ -250,11 +250,15 @@
...
     context 'End of cycle' do
       before do
-        deactivate_feature(:display_apply_button)
         deactivate_feature(:ucas_only_locations)
+        Timecop.freeze(Time.zone.local(2021, 9, 20, 19, 0, 0))
         visit course_path(course.provider_code, course.course_code)
       end
 
+      after do
+        Timecop.return
```
Comment (duncanjbrown) [inline spec/features/courses/show_spec.rb:259]: Could this be done with an `around`?

File: spec/requests/cycle_ending_soon_spec.rb [hunk 7]
```diff
@@ -10,21 +10,13 @@
   end
 
   context 'cycle ending soon' do
-    before do
```
Comment (duncanjbrown) [inline spec/requests/cycle_ending_soon_spec.rb:13]: 🎉

**Review by duncanjbrown — APPROVED** [review]
Looks good — optional refactor in comment!

File: app/components/deadline_banner_component.html.erb (hunk 1 above)
Comment (davidgisbey) [reply app/components/deadline_banner_component.html.erb:12]: Yea i'm going to add a proper dates helper like we have on Apply but in a separate PR as i'm aware this needs to go out today

**Review by EmmaFrith — APPROVED** [review]
thanks @davidgisbey

**Comment by davidgisbey** [comment]
I'll make the requested changes over the weekend and make sure it's shipped by Sunday night!

**Pull request merged**